### PR TITLE
Support Elasticsearch native users

### PR DESCRIPTION
**Pull request by Suraiya-Hameed**
## Description

- Create a new ConfigMap `tigera-known-oidc-users` and a Secrete `tigera-known-oidc-users-credentials`, if external idP is configured and elasticsearch uses basic license. This ConfigMap will be used by tigera-manager to add end-user information needed to create elasticsearch native users.
- Delete ConfigMap `tigera-known-oidc-users` and the Secrete `tigera-known-oidc-users-credentials`, if external idP is not configured or elasticsearch is not in basic license
- Add permissions to manager and kube-controller to access the new ConfigMap and Secret

## For PR author

- [x] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Brian-McM) [inline pkg/controller/logstorage/logstorage_controller.go:687]: The render package allows you to return objects that need to be either created or deleted, this isn't the way to remove items, the second array in the declaration of `Objects() ([]runtime.Object, []runtime.Object)` is for object to delete.

The correct pattern for this (correct as in how the operator team decided to do this) would be look up the config map and secret and pass them to the render code. The render code can then decide whether or not to create the object or delete it. If these object don't exist then you would be passing nil to the render code, in which case the render code can take that as a sign to initialise those config maps.

**Review by Brian-McM — CHANGES_REQUESTED** [review]

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Brian-McM) [inline pkg/controller/logstorage/logstorage_controller.go:470]: Could you please follow the pattern that the other components use and pass the component to render and let it decide what to do if the component exists or not.

For instance, the config map might exist but in future iterations there may be upgrades that we have to make to the config map or changes to make, like adding annotations or labels or completely restructuring the resource.

The general pattern has been to grab a resource, if it exists re render it anyway because you may have updates to it or somebody may have changed it outside of the operator and we want to reconcile it. With the config map and secret we do want the Data edited, but the operator can make sure that config map a secret meta data that should not be changed is always accurate.

Also to note, if the data is changed while the operator is reconciling the request to update the config map will fail and be re attempted later because the revision version will be different, so this is a safe pattern.

File: pkg/render/kube-controllers.go
Comment (Brian-McM) [inline pkg/render/kube-controllers.go:201]: After chatting about this with @doublek we may need to make sure that we're restricting access appropriate, so that kube controllers can only access these resources in the tigera-elasticsearch namespace.

I think you're going to need to create a regular role in the tigera-elasticsearch namespace and bind kube controllers to that. Please give that a though and start a conversation with @doublek and I.

File: pkg/render/logstorage.go
Comment (Brian-McM) [inline pkg/render/logstorage.go:334]: I added comments in the controller about this, we should be passing the config map and secrets to render to be re rendered so that in the future if there's changes we need to make to any meta data it will be easily applied, and there's no worry about overriding the config map data.

**Review by Brian-McM — CHANGES_REQUESTED** [review]

File: pkg/render/kube-controllers.go
Comment (Suraiya-Hameed) [reply pkg/render/kube-controllers.go:201]: yes, creating a roles in tigera-elasticsearch namespace and binding it to manager and kube-controller is definitely better option than clusterrole.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [inline pkg/controller/logstorage/logstorage_controller.go:471]: Is there a watch for this resource?

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [inline pkg/controller/logstorage/logstorage_controller.go:472]: Same here.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [inline pkg/controller/logstorage/logstorage_controller.go:664]: I think the name should just be `getOIDCUserConfigMap`.
Also I think it should return an error. We need to be able to tell if the error was because the resource doesn't exist or there was an error with the query.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [inline pkg/controller/logstorage/logstorage_controller.go:678]: Same comments on this function.

**Review by tmjd — CHANGES_REQUESTED** [review]

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Suraiya-Hameed) [reply pkg/controller/logstorage/logstorage_controller.go:471]: There is no watch on this ConfigMap or Secret, they should be created or deleted depending on es license type and dex config, we already have watch on them, also data in this ConfigMap & Secret will be updated by other components (es-proxy and kube-controller) so we don't have to watch for changes.

Comment (tmjd) [reply pkg/controller/logstorage/logstorage_controller.go]: I think we still need watches for them.
Unless it is guaranteed that the ConfigMap and Secret will be updated before the license type and dex config are updated. If not then it would be possible that the controller gets an update (from the watch) for ES license type and dex config, logstorage controller Reconciles, then the ConfigMap and Secret are updated. Logstorage would have missed the updates to the ConfigMap and Secret.

**Review by tmjd — APPROVED** [review]
LGTM

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Review by Brian-McM — APPROVED** [review]

**Pull request merged**
